File: 33-search-in-rotated-sorted-array/search-in-rotated-sorted-array.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        n = len(nums)
        l, r = 0, n-1
        while l>-1 and r<n and l <= r :
          m = (l+r)//2
          mid = nums[m]

          if l==r and mid!=target:
            break

          print(f"checking: ({m} in [{l}, {r}]) -> [{nums[l]},{nums[r]}], mid: {mid}")
          if mid==target:
            return m
          
          if nums[l] < nums[r]:
            print('sorted case, doing normal bs')
            if target > mid:
              print('sorted case, going to right half next')
              l = m+1
            else:
              print('sorted case, going to left half next')
              r = m-1
          elif nums[l] <= mid:
            print('rotated case: mid in left subarray')
            if target < mid:
              if target >= nums[l]:
                r = m-1
              else:
                l = m+1
            else:
              l = m+1
          else:
            print('rotated case: mid in right subarray')
            if target < mid:
              r = m-1
            elif target > nums[r]:
              r = m-1
            else:
              l = m+1
        return -1
            

File: 33-search-in-rotated-sorted-array/search-in-rotated-sorted-array.java
class Solution {
    public int search(int[] a, int target) {
        int l=0, r=a.length-1;
        while(l<=r){
            int m = l + (r-l)/2;

            if(a[m] == target) return m;
            if(a[l] == target) return l;
            if(a[r] == target) return r;

            if(a[l] < a[r]){
                if(a[m] > target) r = m - 1;
                else l = m + 1;
            }else if(a[m] > a[l]){
                //left sub array
                if(target > a[l] && target < a[m]){
                    r = m - 1; //find in current region only
                }else{
                    l = m + 1; //move to other region
                }
            }else{
                //right sub array
                if (target < a[r] && target > a[m]){
                    l = m + 1; //find in current region
                }else{
                    r = m - 1; //move to other region
                }
            }
        }
        return -1;  
    }
}